compiler/qsc_data_structures/src/span.rs

```diff
@@ -74,3 +74,18 @@ impl From<Span> for SourceSpan {
         Self::from((value.lo as usize)..(value.hi as usize))
     }
 }
+
+#[allow(clippy::module_name_repetitions)]
+pub trait WithSpan {
+    #[must_use]
+    fn with_span(self, span: Span) -> Self;
+}
+
+impl<T> WithSpan for Box<T>
+where
+    T: WithSpan,
+{
+    fn with_span(self, span: Span) -> Self {
+        Box::new((*self).with_span(span))
+    }
```

Comment:
Super nit: I think you can avoid a reallocation this way:
```suggestion
    fn with_span(mut self, span: Span) -> Self {
        *self = (*self).with_span(span);
        self
    }
```
You should check my work though, I'm like 90% sure.

Reply:
I think this adds a requirement that the objects using this method must be mutable, which is not desirable.

Reply:
The other option I could explore here is to not use `Default` at all (I know, flies in the face of my PR name). If I instead introduce a new trait called something like `Error` or `AsError` then I can have the option to pass the span as an argument to the contruction of the error instance. And likewise implement it as a kind of constructor here.

Reply:
Ok! I think I found a solution, which ironically involved adding one more trait: `DefaultWithSpan`. I was able to implement that in a way that avoids the extra allocation for `Box` and doesn't require special extra implementation for any of the other types. Turns out only `qir_ast::Ty` was used directly by `seq` without a box, so just that one implementation was needed after imlementing it for `Box`.

Reply:
> I think this adds a requirement that the objects using this method must be mutable, which is not desirable.

@ScottCarda-MS  `self` and `mut self` both consume `self`, so it shouldn't affect anything from the caller's perspective. The `mut` only matters to the implementation - it's there so that you can assign to `*self` (which you already own) without the compiler yelling at you. (I did find this a bit perplexing at first. I couldn't find a good documentation link for it, but the OP in this thread sums up my reaction to it. https://users.rust-lang.org/t/whats-the-point-of-arguments-with-mutable-ownership/23315)

@swernli did `DefaultWithSpan` have to be a trait? Why not have a standalone function like

```rs
fn default_with_span<T>(span: Span) -> T where T: Default + WithSpan {
    Box::new(T::default().with_span(span))
}

// call site
default_with_span::<T>(span);
```

Is one of those more idiomatic than the other?


I know the PR is merged (yay!) but just responding for the sake of closure.

Reply:
Oh, I misunderstood what you were suggesting before, and you are right, this did not need to be a trait. The impl for `Box<T>` can have `mut self` in the signature without requiring the signature for the trait itself to change. I've put this up in a small PR #852. Thanks for following up and asking more about this!